### ISSUE TITLE
Multiple 'buf' buffer in afl-mutations.h

### ISSUE THREAD

**Issue by futhewo** [AFLplusplus/AFLplusplus#1931]
Hello everyone, thanks for the tool.

I am currently developping a custom mutator that calls afl_mutate on part of its data.
I get a "Buffer overflow detected" bug while calling afl_mutate.
After investigation, it occurs in the last memcpy of the MUT_ASCIINUM case (line 2482 in current dev branch).
`len += (new_len - old_len);
memcpy(buf, tmp_buf, len);
`
It comes from the fact that there is two 'buf' in this function:

- One that comes as argument.
- One that is defined inside this case and has a size of 20.

This memcpy should write in the former but writes in the latter.
So, when len is above 20, there is indeed a buffer overflow.

I think there is the same problem in MUT_INSERTASCIINUM.

The solution is to rename the temporary in-the-case 'buf' as 'case_buf'. It implies some changes in the memcpys.
Something like that I think:

For MUT_ASCIINUM:
```
        char case_buf[20];
        snprintf(case_buf, sizeof(case_buf), "%" PRId64, val);
        u32 old_len = off2 - off;
        u32 new_len = strlen(case_buf);

        if (old_len == new_len) {

          memcpy(buf + off, case_buf, new_len);

        } else {

          /* Head */

          memcpy(tmp_buf, buf, off);

          /* Inserted part */

          memcpy(tmp_buf + off, case_buf, new_len);

          /* Tail */
          memcpy(tmp_buf + off + new_len, buf + off2, len - off2);

          len += (new_len - old_len);
          memcpy(buf, tmp_buf, len);

        }
```

For MUT_INSERTASCIINUM:
```
        char case_buf[20];
        snprintf(case_buf, sizeof(case_buf), "%llu", val);
        memcpy(buf + pos, case_buf, len);
```

Best regards

**Comment by vanhauser-thc** [comment]
oh, thank you for that!
I made the same mistake in MUT_INSERTASCIINUM, not resulting in a memory corruption, but in a null op.
Not sure why, but I never saw a crash, nor in any of the experiments.
fixed in 37505928bcec63a08fe50cdebdbf7b9b28b952d0

**Comment by futhewo** [comment]
My pleasure.
Thanks for the fix!
